runtime/ints.c

```diff
@@ -139,6 +139,30 @@ CAMLprim value caml_int_of_string(value s)
     return Val_long(parse_intnat(s, 8 * sizeof(value) - 1, INT_ERRMSG));
 }
 
+CAMLprim intnat caml_int_clz(value v)
+{
+#if defined(__has_builtin) && __has_builtin(__builtin_clzl)
+  if (sizeof(v) == sizeof(long))
+    /* v is never zero due to the tagging bit. */
+    return __builtin_clzl(v);
```

Comment:
I don't think that L147 does the right thing for 32-bit targets. You might need something like this:
```
#ifdef ARCH_SIXTYFOUR
return int64_clz(v)
#else
return int32_clz(v)
#endif
```
as defined [here](https://github.com/janestreet/ocaml_intrinsics/blob/98af7a639c1b4a50c9e4b04e9b070a28129ca6a9/src/int_stubs.c#L7-L20).

Reply:
That's why I added the `sizeof(v) == sizeof(long)` check. I do think the code is correct as long as Ocaml uses all bits in `value` (or `v`), so that a (tagged) zero integer always has `sizeof(v) * 8 - 1` zeros. I think that's true for all Ocaml (C) targets, 32-bit or 64-bit.

In the unlike situation that `sizeof(v) != sizeof(long)`, the C implementation that follows still does the right thing (as long as `sizeof(uintnat) == sizeof(value)`).

Reply:
great, thanks for the explanation.